src/idiomatic/leveraging-the-type-system/typestate-pattern/typestate-generics.md

```diff
@@ -0,0 +1,268 @@
+        # Serializer {
+        #     indent: self.indent + 1,
+        #     buffer: self.buffer,
+        #     state: Struct(self.state),
+        # }
+    }
+
+    fn serialize_string(mut self, value: &str) -> Self {
+        // [...]
+        # writeln!(self.buffer, "{}{value},", " ".repeat(self.indent * 2)).unwrap();
+        # self
+    }
+
+    fn finish_list(mut self) -> Serializer<S> {
+        // [...]
+        # self.indent -= 1;
+        # writeln!(self.buffer, "{}]", " ".repeat(self.indent * 2)).unwrap();
+        # Serializer {
+        #     indent: self.indent,
+        #     buffer: self.buffer,
+        #     state: self.state.0,
+        # }
+    }
+}
+
+fn main() {
+    # #[rustfmt::skip]
+    let serializer = Serializer::new()
+        .serialize_struct("Foo")
+            .serialize_property("bar")
+            .serialize_struct("Bar")
+                .serialize_property("baz")
+                .serialize_list()
+                    .serialize_string("abc")
+                    .serialize_struct("Baz")
+                        .serialize_property("partial")
+                        .serialize_string("def")
+                        .serialize_property("empty")
+                        .serialize_struct("Empty")
+                        .finish_struct()
+                    .finish_struct()
+                .finish_list()
+            .finish_struct()
+        .finish_struct();
+
+    # let buffer_size = serializer.buffer_size();
+    let output = serializer.finish();
+
+    # println!("buffer size = {buffer_size}\n---");
+    println!("{output}");
+
+    // These will all fail at compile time:
+
+    // Serializer::new().serialize_list();
+    // Serializer::new().serialize_string("foo");
+    // Serializer::new().serialize_struct("Foo").serialize_string("bar");
+    // Serializer::new().serialize_struct("Foo").serialize_list();
+    // Serializer::new().serialize_property("foo");
+}
+```
+
+<details>
+
+- The full code for this example is available
+  [in the playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=48b106089ca600453f3ed00a0a31af26).
+
+- By using generics to track the parent context, we can construct arbitrarily
+  nested serializers that enforce valid transitions between struct, list, and
+  property states.
+
+- This lets us build a recursive structure while preserving control over what
+  methods are accessible in each state.
+
+- Here's how the flow maps to a state machine:
+
+```bob
+    +-----------+   +---------+------------+-----+
+    |           |   |         |            |     |
+    V           |   V         |            V     |
+                +                                |
+serializer --> structure --> property --> list +-+
+
+    |           |   ^           |          ^
+    V           |   |           |          |
+                |   +-----------+          |
+  String        |                          |
+                +--------------------------+
+```
```

Comment:
Why is there an arrow from `structure` to `list` in this diagram? `Serializer<Struct>` only allows starting a property or finishing the struct.

Reply:
Correct. It's because of your last option. Finishing a struct goes to its parent which can be:

- the root serializer
- another struct
- a list

As such it can also go back to a `list`.